### ISSUE TITLE
LS means with offset option

### ISSUE THREAD

**Issue by elong0527** [rvlenth/emmeans#18]
I am using `lsmeans` and just notice the new package `emmeans`. Hopefully my issue for `lsmeans` has been addressed in `emmeans`.

In lsemans, while we have a model with offset term, the LS means will automatically add the offset term.

It would be nice to have some information in the help page, as this is inconsistent with SAS. (SAS did not add offset term)

It will be nice to have LS means for zero-inflated models that used in pscl::zeroinfl. 

The current code can directly be used. my hack is as below

```
# negtive binomial model for event rate
  fit <- MASS::glm.nb(cnt ~ TRT01P + ... + offset(logT), data = db_cnt)
  .ls <- lsmeans(fit, "TRT")
  .er <- exp(fit_ls[,2] - .ls@grid$.offset.)

# Zero inflated negtive binomial model for event rate
  fit<- pscl::zeroinfl(cnt ~ TRT + ... + offset(logT) | TRT,
                 dist = "negbin", data = db_cnt)  
  .er <- as.numeric(exp(.ls@linfct %*% fit$coefficients$count ))
```

**Comment by rvlenth** [comment]
First, I have not changed anything in the way that **emmeans** handles offsets. I will consider whether I need to clarify the documentation on that matter. However, it does already make clear that EMMs (and LS means) are based on model predictions; and so far as I know, the `predict` methods in any R package that supports models with offsets will add the offset to the predictions. So I think I am being consistent with R, and this seems more important than being consistent with SAS. If SAS does not add offsets to LS means, I would consider that a design defect.

(However, what I think SAS does is to omit the *zero-inflated part*; not the offset. See the [SAS documentation](https://support.sas.com/documentation/cdl/en/statug/63347/HTML/default/viewer.htm#statug_genmod_sect024.htm). That is an option in my setup -- see below.)

I'm confused about what you think is lacking in support for `pscl::zeroinf` objects. If you will look at the documentation in the ["models" vignette](https://cran.r-project.org/web/packages/emmeans/vignettes/models.html), in particular section "C", you will see that there are four options for `mode` (`"response", "count", "zero"`, and `"prob0"`), and two for `lin.pred` (`FALSE` and `TRUE`). These correspond to options in the `predict` method for these objects. The same options were available in the **lsmeans** package -- see `? models`.

If you don't want the offset included, there is a simple way to hack around it:
```
emm <- emmeans( ... )
emm@grid$.offset. <- 0
```
I will consider providing a user interface for removing or changing the offset. But really, as a statistician I think the offset is an inseparable part of the model; it is tantamount to including it in the model as a predictor with fixed coefficient of 1.

**Comment by elong0527** [comment]
Thanks for the very detail responses. I should read the vignettes before raising this issue.  

For the offset term, one example is to calculate event rate in a negative binomial model/Poission model. If we need to evaluate the event rate, the offset term should not be included, because the offset term logT is used in the response part. 

```
fit <- MASS::glm.nb(cnt ~ TRT01P + ... + offset(logT), data = db_cnt)
```

**Comment by rvlenth** [comment]
OK, well that might be a situation that warrants a less hacky way of removing the offset then.

**Comment by rvlenth** [comment]
@elong0527 - you closed this issue, but I'm still thinking about it and want to provide a reasonable interface for ignoring the offset. Clearly, the event-rate situation is a compelling case where you don't want to include the offset in the EMM calculations. 

My question is on *when* to provide for this. One option is to provide an argument in `summary.emmGrid` where the offset can be ignored, but could be included in other summaries of the object. Another would be to provide a mechanism in `ref_grid()` itself (typically its arguments are passed via `emmeans()`) so that the offset is ignored once and for all when the reference grid itself is constructed. Do you have an opinion on this?

**Comment by elong0527** [comment]
Hi @rvlenth , thanks for reopening the issue.

I believe you will be the best person to determine which way to implement it. From a user perspective, I appreciate that the `emmeans` function (I assume this is the function most used by a user) will have an option to turn on/off the offset term.

**Comment by rvlenth** [comment]
I understand, and that's what I suspected. So I'll add that option (once I decide what to call it and how to implement it) to `ref_grid()`, which in turn will make it available to `emmeans()`, `emtrends()`, and oters that need to construct a reference grid. 

I guess my first thought is to have `offset` as a numeric argument, so that `offset = 0` would set all the offset values to 0 but allow manually setting the offset(s) to something else as seen appropriate. That allows some interesting extensions. For example, if your underlying reference is rate per day and you want rate per week instead, you could use `emmeans(..., offset = log(7))`. I'll keep you posted.

**Comment by rvlenth** [comment]
I decided to add a (numeric) `offset` argument in two places: 

- in `emmeans()`, we can specify a scalar or vector offset value, which adds an offset, or replaces an existing offset. The offset value is recycled as needed.

- in `ref_grid()`, only a scalar offset is allowed (applies to all combinations)

My thinking here is that if I had only provided `offset` in `ref_grid()`, then it becomes dangerous to pass it vector values because the user may easily misunderstand the order of factor combinations in the grid; whereas this can be anticipated in `emmeans()`.

The changes will be incorporated in the next push to github.
